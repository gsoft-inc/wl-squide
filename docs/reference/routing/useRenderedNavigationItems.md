---
order: 80
toc:
    depth: 2-3
---

# useRenderedNavigationItems

Recursively parse a navigation items structure to transform the items into React Elements.

> The [useNavigationItems](../runtime/useNavigationItems.md) hook returns the navigation items tree structure as is, meaning the consumer has to recursively parse the structure to transform the items into actual React Elements.
>
> As it's a non-trivial process, Squide provides this utility hook.

## Reference

```ts
const elements = useRenderedNavigationItems(
    navigationItems: [],
    renderItem: (item, key, index, level) => {},
    renderSection: (elements, key, index, level) => {})
```

### Parameters

- `navigationItems`: An array of `NavigationLink | NavigationSection` to render.
- `renderItem`: A function to render a link from a navigation item
- `renderSection`: A function to render a section from a collection of items.

#### `NavigationLink`

Accept any properties of a React Router [Link](https://reactrouter.com/en/main/components/link) component with the addition of:

- `$key`: An optional key identifying the link. Usually used as the React element [key](https://legacy.reactjs.org/docs/lists-and-keys.html#keys) property.
- `$label`: The link label. Could either by a `string` or a `ReactNode`.
- `$canRender`: An optional function accepting an object and returning a `boolean` indicating whether or not the link should be rendered.
- `$additionalProps`: An optional object literal of additional props to apply to the link component.

#### `NavigationSection`

- `$key`: An optional key identifying the section. Usually used as the React element [key](https://legacy.reactjs.org/docs/lists-and-keys.html#keys) property.
- `$label`: The section label. Could either by a `string` or a `ReactNode`.
- `$canRender`: An optional function accepting an object and returning a `boolean` indicating whether or not the section should be rendered.
- `$additionalProps`: An optional object literal of additional props to apply to the section component.
- `children`: The section items.

### Returns

An array of `ReactElement`.

## Usage

### Render nested items

!!!info
We recommend always providing a `$key` property for a navigation item, as it ensures the menus doesn't flicker when [deferred registrations](../registration/registerLocalModules.md#defer-the-registration-of-navigation-items) are updated. Be sure to use a unique key.

When no `$key` property is provided, a default key value is computed based on the `index` and `level` properties. While this works in most cases, the default key cannot guarantee that the menu won't flicker during updates.
!!!

```tsx !#38-40,42-48,52 host/src/RootLayout.tsx
import type { ReactNode } from "react";
import { Link, Outlet } from "react-router-dom";
import { 
    useNavigationItems, 
    useRenderedNavigationItems, 
    isNavigationLink,
    type RenderItemFunction, 
    type RenderSectionFunction, 
    type NavigationLinkRenderProps, 
    type NavigationSectionRenderProps
} from "@squide/react-router";

type RenderLinkItemFunction = (item: NavigationLinkRenderProps, key: string) => ReactNode;

type RenderSectionItemFunction = (item: NavigationSectionRenderProps, key: string) => ReactNode;

const renderLinkItem: RenderLinkItemFunction = ({ label, linkProps, additionalProps }, key) => {
    return (
        <li key={key}>
            <Link {...linkProps} {...additionalProps}>
                {label}
            </Link>
        </li>
    );
};

const renderSectionItem: RenderSectionItemFunction = ({ label, section }, key) => {
    return (
        <li key={key}>
            {label}
            <div>
                ({section})
            </div>
        </li>
    );
};

const renderItem: RenderItemFunction = (item, key) => {
    return isNavigationLink(item) ? renderLinkItem(item, key) : renderSectionItem(item, key;
};

const renderSection: RenderSectionFunction = (elements, key, index, level) => {
    return (
        <ul key={key}>
            {elements}
        </ul>
    );
};

export function RootLayout() {
    const navigationItems = useNavigationItems();
    const navigationElements = useRenderedNavigationItems(navigationItems, renderItem, renderSection);

    return (
        <>
            <nav>{navigationElements}</nav>
            <Outlet />
        </>
    );
}
```

### Render dynamic segments

The `to` property of a navigation item can include dynamic segments (`/user-profile/:userId`), enabling the rendering of dynamic routes based on contextual values. To resolve a route dynamic segments, use the [resolveRouteSegments](resolveRouteSegments.md) function.

```tsx !#14,18,21,39-45,56,59 host/src/UserProfileLayout.tsx
import type { ReactNode } from "react";
import { Link, Outlet } from "react-router-dom";
import { 
    useNavigationItems, 
    useRenderedNavigationItems,
    isNavigationLink,
    resolveRouteSegments
    type RenderItemFunction, 
    type RenderSectionFunction, 
    type NavigationLinkRenderProps, 
    type NavigationSectionRenderProps
} from "@squide/react-router";

type RenderLinkItemFunction = (item: NavigationLinkRenderProps, key: string, userId: string) => ReactNode;

type RenderSectionItemFunction = (item: NavigationSectionRenderProps, key: string) => ReactNode;

const renderLinkItem: RenderLinkItemFunction = ({ label, { to, ...linkProps}, additionalProps }, key, userId) => {
    return (
        <li key={key}>
            <Link to={resolveRouteSegments(to as string, { userId })} {...linkProps} {...additionalProps}>
                {label}
            </Link>
        </li>
    );
};

const renderSectionItem: RenderSectionItemFunction = ({ label, section }, key) => {
    return (
        <li key={key}>
            {label}
            <div>
                ({section})
            </div>
        </li>
    );
};

function renderItem(userId: string) {
    const fct: RenderItemFunction = (item, key) => {
        return isNavigationLink(item) ? renderLinkItem(item, key, userId) : renderSectionItem(item, key);
    };

    return fct;
}

const renderSection: RenderSectionFunction = (elements, key) => {
    return (
        <ul key={key}>
            {elements}
        </ul>
    );
};

export function UserProfileLayout() {
    const { userId } = useParams();

    const navigationItems = useNavigationItems({ menuId: "/user-profile" });
    const navigationElements = useRenderedNavigationItems(navigationItems, renderItem(userId), renderSection);

    return (
        <>
            <nav>{navigationElements}</nav>
            <Outlet />
        </>
    );
}
```

```tsx !#7 remote-module/src/register.tsx
import type { ModuleRegisterFunction, FireflyRuntime } from "@squide/firefly";

export const register: ModuleRegisterFunction<FireflyRuntime> = (runtime) => {
    runtime.registerNavigationItem({
        $key: "user-profile",
        $label: "User profile",
        to: "/user-profile/:userId"
    }, {
        menuId: "/user-profile"
    });
}
```


