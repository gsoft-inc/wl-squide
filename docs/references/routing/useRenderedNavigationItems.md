# useRenderedNavigationItems

Recursively parse a navigation items structure to transform the items into React Elements.

> The `useNavigationItems` hook returns the navigation items tree structure as is, meaning the consumer has to recursively parse the structure to transform the items into actual React Elements.
>
> As it's a non-trivial process, the shell provides this utility hook to help with that.

## Reference

```ts
const elements = useRenderedNavigationItems(navigationItems: [], renderItem: () => {}, renderSection: () => {})
```

### Parameters

- `navigationItems`: An array of `NavigationItem` to render.
- `renderItem`: A function to render a single link from a navigation item
- `renderSection`: A function to render a section from a collection of items.

### Returns

An array of `ReactElement`.

## Usage

```tsx !#38-40,42-48,52 host/src/RootLayout.tsx
import type { ReactNode } from "react";
import { Link, Outlet } from "react-router-dom";
import { 
    useNavigationItems, 
    useRenderedNavigationItems, 
    isNavigationLink,
    type RenderItemFunction, 
    type RenderSectionFunction, 
    type NavigationLinkRenderProps, 
    type NavigationSectionRenderProps
} from "@squide/react-router";

type RenderLinkItemFunction = (item: NavigationLinkRenderProps, index: number, level: number) => ReactNode;

type RenderSectionItemFunction = (item: NavigationSectionRenderProps, index: number, level: number) => ReactNode;

const renderLinkItem: RenderLinkItemFunction = ({ label, linkProps, additionalProps }, index, level) => {
    return (
        <li key={`${level}-${index}`}>
            <Link {...linkProps} {...additionalProps}>
                {label}
            </Link>
        </li>
    );
};

const renderSectionItem: RenderSectionItemFunction = ({ label, section }, index, level) => {
    return (
        <li key={`${level}-${index}`}>
            {label}
            <div>
                ({section})
            </div>
        </li>
    );
};

const renderItem: RenderItemFunction = (item, index, level) => {
    return isNavigationLink(item) ? renderLinkItem(item, index, level) : renderSectionItem(item, index, level);
};

const renderSection: RenderSectionFunction = (elements, index, level) => {
    return (
        <ul key={`${level}-${index}`}>
            {elements}
        </ul>
    );
};

export default function RootLayout() {
    const navigationItems = useNavigationItems();
    const navigationElements = useRenderedNavigationItems(navigationItems, renderItem, renderSection);

    return (
        <>
            <nav>
                {navigationElements}
            </nav>
            <Outlet />
        </>
    );
}
```
