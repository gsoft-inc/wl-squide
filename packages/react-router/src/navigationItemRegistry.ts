import { isNil } from "@squide/core";
import type { ReactNode } from "react";
import type { LinkProps, To } from "react-router-dom";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ToFunction = (...args: any[]) => To;

export interface NavigationLink extends Omit<LinkProps, "to" | "children"> {
    $label: ReactNode;
    $to: To | ToFunction;
    $additionalProps?: Record<string, unknown>;
    children?: never;
}

export interface NavigationSection {
    $label: ReactNode;
    $additionalProps?: Record<string, unknown>;
    children: NavigationItem[];
    $to?: never;
}

export type NavigationItem = NavigationLink | NavigationSection;

// Will be exposed externally but is only expected to be used internally.
export function isLinkItem(item: NavigationItem): item is NavigationLink {
    return !isNil((item as NavigationLink).$to);
}

export type RootNavigationItem = NavigationItem & {
    // Highest priority is rendered first.
    $priority?: number;
};

export class NavigationItemRegistry {
    readonly #menus: Map<string, RootNavigationItem[]> = new Map();

    add(menuId: string, navigationItem: RootNavigationItem) {
        // Create a new array so the navigation items array is immutable.
        const items = [
            ...(this.#menus.get(menuId) ?? []),
            navigationItem
        ];

        this.#menus.set(menuId, items);
    }

    getItems(menuId: string) {
        return this.#menus.get(menuId);
    }
}
