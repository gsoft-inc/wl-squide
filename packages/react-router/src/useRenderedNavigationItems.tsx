import { useMemo, type ReactNode } from "react";
import { isLinkItem, type NavigationItem, type NavigationLink, type NavigationSection, type RootNavigationItem } from "./navigationItemRegistry.ts";

import { isNil } from "@squide/core";
import type { LinkProps } from "react-router-dom";

export interface NavigationLinkRenderProps {
    label: ReactNode;
    linkProps: Omit<LinkProps, "children">;
    additionalProps: Record<string, unknown>;
}

export interface NavigationSectionRenderProps {
    label: ReactNode;
    section: ReactNode;
    additionalProps: Record<string, unknown>;
}

export type NavigationItemRenderProps = NavigationLinkRenderProps | NavigationSectionRenderProps;

export function isNavigationLink(item: NavigationItemRenderProps): item is NavigationLinkRenderProps {
    return !isNil((item as NavigationLinkRenderProps).linkProps);
}

export type RenderItemFunction = (item: NavigationItemRenderProps, index: number, level: number) => ReactNode;

export type RenderSectionFunction = (elements: ReactNode[], index: number, level: number) => ReactNode;

function toLinkProps({ label, additionalProps, ...linkProps }: NavigationLink): NavigationLinkRenderProps {
    return {
        label,
        linkProps,
        additionalProps: additionalProps ?? {}
    };
}

function toMenuProps({ label, additionalProps }: NavigationSection, sectionElement: ReactNode): NavigationSectionRenderProps {
    return {
        label,
        section: sectionElement,
        additionalProps: additionalProps ?? {}
    };
}

function renderItems(items: NavigationItem[], renderItem: RenderItemFunction, renderSection: RenderSectionFunction, index: number, level: number) {
    const itemElements = items.map((x, itemIndex) => {
        let itemElement: ReactNode;

        if (isLinkItem(x)) {
            itemElement = renderItem(toLinkProps(x), itemIndex, level);
        } else {
            const sectionElement = renderItems(x.children, renderItem, renderSection, 0, level + 1);

            itemElement = renderItem(toMenuProps(x, sectionElement), itemIndex, level);
        }

        return itemElement;
    });

    return renderSection(itemElements, index, level);
}

export function useRenderedNavigationItems(
    navigationItems: RootNavigationItem[],
    renderItem: RenderItemFunction,
    renderSection: RenderSectionFunction
) {
    return useMemo(() => {
        // Highest priority is rendered first.
        const sortedItems = [...navigationItems]
            .sort((x, y) => {
                // Default an item priority to 0 to support negative priority.
                const xp = x.priority ?? 0;
                const yp = y.priority ?? 0;

                if (xp === yp) {
                    return 0;
                }

                return xp! > yp! ? -1 : 1;
            })
            // priority is intentionally omitted.
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            .map(({ priority, ...itemProps }) => itemProps);

        return renderItems(sortedItems, renderItem, renderSection, 0, 0);
    }, [navigationItems, renderItem, renderSection]);
}
