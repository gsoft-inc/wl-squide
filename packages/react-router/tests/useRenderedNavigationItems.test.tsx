import renderer from "react-test-renderer";
import { type RootNavigationItem } from "../src/navigationItemRegistry.ts";
import { useRenderedNavigationItems, type RenderItemFunction, type RenderSectionFunction, isNavigationLink, type NavigationLinkRenderProps, type NavigationSectionRenderProps } from "../src/useRenderedNavigationItems.tsx";
import { useCallback, type ReactElement } from "react";
import { renderHook } from "@testing-library/react";

type RenderLinkItemFunction = (item: NavigationLinkRenderProps, index: number, level: number) => ReactElement;

type RenderSectionItemFunction = (item: NavigationSectionRenderProps, index: number, level: number) => ReactElement;

interface TestComponentProps {
    navigationItems: RootNavigationItem[];
}

// Not the prettiest mock but it's easier than simpler than using createMemoryRouter and
// it provides the required level of testing when combined with snapshot tests.
function Link(props: Record<string, unknown>) {
    return (
        <div {...props} />
    );
}

function TestComponent({ navigationItems }: TestComponentProps) {
    const renderLink: RenderLinkItemFunction = useCallback(({ label, linkProps, additionalProps }, index, level) => {
        return (
            <li key={`${level}-${index}`} {...additionalProps}>
                <Link {...linkProps}>
                    {label}
                </Link>
            </li>
        );
    }, []);

    const renderMenu: RenderSectionItemFunction = useCallback(({ label, section, additionalProps }, index, level) => {
        return (
            <li key={`${level}-${index}`} {...additionalProps}>
                {label}
                {section}
            </li>
        );
    }, []);

    const renderItem: RenderItemFunction = useCallback((item, index, level) => {
        return isNavigationLink(item) ? renderLink(item, index, level) : renderMenu(item, index, level);
    }, [renderLink, renderMenu]);

    const renderSection: RenderSectionFunction = useCallback((elements, index, level) => {
        return (
            <ul key={`${level}-${index}`}>
                {elements}
            </ul>
        );
    }, []);

    // eslint-disable-next-line testing-library/render-result-naming-convention
    return useRenderedNavigationItems(navigationItems, renderItem, renderSection);
}

test("highest priority goes first", () => {
    const navigationItems: RootNavigationItem[] = [
        {
            to: "/foo",
            label: "Foo"
        },
        {
            to: "/bar",
            label: "Bar",
            priority: 5
        },
        {
            to: "/toto",
            label: "Toto",
            priority: 99
        },
        {
            to: "/tutu",
            label: "Tutu"
        }
    ];

    const tree = renderer
        .create(<TestComponent navigationItems={navigationItems} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("support 2 section levels", () => {
    const navigationItems: RootNavigationItem[] = [
        {
            to: "/foo",
            label: "Foo"
        },
        {
            label: "Bar",
            children: [
                {
                    to: "/toto",
                    label: "Toto"
                },
                {
                    to: "/tutu",
                    label: "Tutu"
                }
            ]
        }
    ];

    const tree = renderer
        .create(<TestComponent navigationItems={navigationItems} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("support 3 section levels", () => {
    const navigationItems: RootNavigationItem[] = [
        {
            to: "/foo",
            label: "Foo"
        },
        {
            label: "Bar",
            children: [
                {
                    to: "/toto",
                    label: "Toto"
                },
                {
                    label: "Tutu",
                    children: [
                        {
                            to: "/titi",
                            label: "Titi"
                        }
                    ]
                }
            ]
        }
    ];

    const tree = renderer
        .create(<TestComponent navigationItems={navigationItems} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("Link item additionalProps are rendered", () => {
    const navigationItems: RootNavigationItem[] = [
        {
            to: "/foo",
            label: "Foo",
            additionalProps: {
                style: { color: "red" }
            }
        },
        {
            to: "/bar",
            label: "Bar"
        }
    ];

    const tree = renderer
        .create(<TestComponent navigationItems={navigationItems} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("Section item additionalProps are rendered", () => {
    const navigationItems: RootNavigationItem[] = [
        {
            label: "Foo",
            children: [
                {
                    to: "/bar",
                    label: "Bar"
                }
            ],
            additionalProps: {
                style: { color: "red" }
            }
        }
    ];

    const tree = renderer
        .create(<TestComponent navigationItems={navigationItems} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("doesn't rerender when the navigation items haven't changed", () => {
    const initialItems = [
        {
            to: "/foo",
            label: "Foo"
        }
    ];

    const renderItem = jest.fn(() => <div>Item</div>);
    const renderSection = jest.fn(() => <div>Section</div>);

    const { rerender } = renderHook(({ navigationItems: x }) => useRenderedNavigationItems(x, renderItem, renderSection), {
        initialProps: {
            navigationItems: initialItems
        }
    });

    rerender({
        navigationItems: initialItems
    });

    expect(renderItem).toHaveBeenCalledTimes(1);
    expect(renderSection).toHaveBeenCalledTimes(1);
});

test("rerender when the navigation items change", () => {
    const initialItems = [
        {
            to: "/foo",
            label: "Foo"
        }
    ];

    const renderItem = jest.fn(() => <div>Item</div>);
    const renderSection = jest.fn(() => <div>Section</div>);

    const { rerender } = renderHook(({ navigationItems: x }) => useRenderedNavigationItems(x, renderItem, renderSection), {
        initialProps: {
            navigationItems: initialItems
        }
    });

    rerender({
        navigationItems: [
            {
                to: "/bar",
                label: "Bar"
            }
        ]
    });

    expect(renderItem).toHaveBeenCalledTimes(2);
    expect(renderSection).toHaveBeenCalledTimes(2);
});
