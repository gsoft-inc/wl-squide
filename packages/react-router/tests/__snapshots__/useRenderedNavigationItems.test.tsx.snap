// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link item additionalProps are rendered 1`] = `
<ul>
  <li
    style={
      {
        "color": "red",
      }
    }
  >
    <div
      to="/foo"
    >
      Foo
    </div>
  </li>
  <li>
    <div
      to="/bar"
    >
      Bar
    </div>
  </li>
</ul>
`;

exports[`Section item additionalProps are rendered 1`] = `
<ul>
  <li
    style={
      {
        "color": "red",
      }
    }
  >
    Foo
    <ul>
      <li>
        <div
          to="/bar"
        >
          Bar
        </div>
      </li>
    </ul>
  </li>
</ul>
`;

exports[`highest priority goes first 1`] = `
<ul>
  <li>
    <div
      to="/toto"
    >
      Toto
    </div>
  </li>
  <li>
    <div
      to="/bar"
    >
      Bar
    </div>
  </li>
  <li>
    <div
      to="/foo"
    >
      Foo
    </div>
  </li>
  <li>
    <div
      to="/tutu"
    >
      Tutu
    </div>
  </li>
</ul>
`;

exports[`negative priority goes last 1`] = `
<ul>
  <li>
    <div
      to="/foo"
    >
      Foo
    </div>
  </li>
  <li>
    <div
      to="/bar"
    >
      Bar
    </div>
  </li>
  <li>
    <div
      to="/tutu"
    >
      Tutu
    </div>
  </li>
  <li>
    <div
      to="/toto"
    >
      Toto
    </div>
  </li>
</ul>
`;

exports[`support 2 section levels 1`] = `
<ul>
  <li>
    <div
      to="/foo"
    >
      Foo
    </div>
  </li>
  <li>
    Bar
    <ul>
      <li>
        <div
          to="/toto"
        >
          Toto
        </div>
      </li>
      <li>
        <div
          to="/tutu"
        >
          Tutu
        </div>
      </li>
    </ul>
  </li>
</ul>
`;

exports[`support 3 section levels 1`] = `
<ul>
  <li>
    <div
      to="/foo"
    >
      Foo
    </div>
  </li>
  <li>
    Bar
    <ul>
      <li>
        <div
          to="/toto"
        >
          Toto
        </div>
      </li>
      <li>
        Tutu
        <ul>
          <li>
            <div
              to="/titi"
            >
              Titi
            </div>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
`;
